### PR TITLE
feat(`monitoringc`): initialize `MsgCreateClient`

### PR DESCRIPTION
**Pull request by lubtd**
## Feature

- Initialize the `MsgCreateClient` message. The validator set hash is verified, the content with `launch` module will be verified in https://github.com/tendermint/spn/issues/436
- Implement the `ibctypes` package for objects related to the IBC handshake. The package provides utility methods to parse dumped Validator Set and Consensus State from the chain CLI
- Add the list of verified IBC Clients
- Add a query for the list of verified IBC Clients depending on the chain ID

## Further steps:
#446
#447
#448
#449

## Testing

https://github.com/lubtd/orbit can be used for testing. The chain incorporate the Monitoring Provider module and is served on another port than SPN

1 - serve both chains
```
starport c serve -r
```

2 - create a chain on SPN
```
spnd tx profile create-coordinator --from alice
spnd tx launch create-chain orbit-1 orbit.com 0xaaa --from alice
```

3 - fetch consensus state and validator set in files at height 1 on orbit
```
orbitd q tendermint-validator-set 1 --node "tcp://localhost:26659" > vs.yaml
orbitd q ibc client self-consensus-state --height 1 --node "tcp://localhost:26659" > cs.yaml
```

4 - Create the IBC client on SPN for the Testnet with the dumped data
```
spnd tx monitoring-consumer create-client 1 cs.yaml vs.yaml --from alice
```

5 - Check the verified IBC Client has been created
```
spnd q monitoring-consumer verified-client-ids 1
```

### CODE REVIEW

File: x/monitoringc/types/errors.go [hunk 1]
```diff
@@ -8,7 +8,11 @@ import (
 
 // x/monitoringc module sentinel errors
 var (
-	ErrSample               = sdkerrors.Register(ModuleName, 1100, "sample error")
-	ErrInvalidPacketTimeout = sdkerrors.Register(ModuleName, 1500, "invalid packet timeout")
-	ErrInvalidVersion       = sdkerrors.Register(ModuleName, 1501, "invalid version")
+	ErrInvalidPacketTimeout    = sdkerrors.Register(ModuleName, 1, "invalid packet timeout")
```
Comment (ivanovpetr) [inline x/monitoringc/types/errors.go:11]: As I know code 1 is reserved for internal usage. Error codes declaration must start with 2.

**Review by Pantani — APPROVED** [review]
CLI works well!!!

I have a doubt. Where are the generated proto files coming from?

File: proto/types/ibctypes.proto [hunk 2]
```diff
@@ -0,0 +1,39 @@
+syntax = "proto3";
+package tendermint.spn.ibctypes;
+
+import "gogoproto/gogo.proto";
+
+option go_package = "github.com/tendermint/spn/pkg/ibctypes";
+
+// MerkeRool represents a Merkel Root in ConsensusState
+message MerkeRool {
+    string hash = 1 [(gogoproto.moretags) = "yaml:\"hash\""];
```
Comment (Pantani) [inline proto/types/ibctypes.proto:10]: @froyd told it some slack thread:

> no, don’t use custom tags.. they’re not protojson compliant
> ...
> they should not be used. they’re there for backwards compatibility but when we will move to new proto codegen those will all disappear

File: pkg/ibctypes/consensus_state.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	err = yaml.Unmarshal(f, &csf)
+	return
+}
+
+// NewConsensusState initializes a new consensus state
+func NewConsensusState(timestamp, nextValHash, rootHash string) ConsensusState {
+	return ConsensusState{
+		NextValHash: nextValHash,
+		Timestamp:   timestamp,
+		Root: MerkeRool{
```
[the remark this answers is not in the document]
Comment (lubtd) [reply pkg/ibctypes/consensus_state.go:38]: Yes thanks for spotting this

File: x/monitoringc/types/errors.go (hunk 1 above)
Comment (lubtd) [reply x/monitoringc/types/errors.go:11]: Good point👍

File: proto/monitoringc/query.proto [hunk 4]
```diff
@@ -27,4 +32,14 @@ message QueryParamsResponse {
...
 }
 
+message QueryVerifiedClientIdsRequest {
+  uint64 launchID = 1;
+cosmos.base.query.v1beta1.PageRequest pagination = 2;
+}
+
+message QueryVerifiedClientIdsResponse {
+  repeated string clientIds = 1;
+cosmos.base.query.v1beta1.PageResponse pagination = 2;
```
[the remark this answers is not in the document]
Comment (lubtd) [reply proto/monitoringc/query.proto:42]: I think it's Starport generation

File: proto/types/ibctypes.proto (hunk 2 above)
Comment (lubtd) [reply proto/types/ibctypes.proto:10]: Ok good to know
Is `[(gogoproto.nullable) = false]` concerned as well?

Comment (Pantani) [reply proto/types/ibctypes.proto]: no, only the custom tags I think

File: pkg/ibctypes/validator_set.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	"gopkg.in/yaml.v2"
+)
+
+const TypeEd25519 = "tendermint/PubKeyEd25519"
+
+// ParseValidatorSetFromFile parses a YAML dumped Validator Set file and returns a new Validator Set
+// TODO: Support other format than YAML if there are other format of dumped file
+func ParseValidatorSetFromFile(filePath string) (vs ValidatorSet, err error) {
+	// parse file
+	var vsf struct {
```
Comment (Pantani) [inline pkg/ibctypes/validator_set.go:23]: maybe we should create a global struct type outside the method for it

File: pkg/ibctypes/validator_set.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+				Type  string `yaml:"type"`
+				Value string `yaml:"value"`
+			} `yaml:"pub_key"`
+		} `yaml:"validators"`
+	}
+	f, err := os.ReadFile(filePath)
+	if err != nil {
+		return vs, err
+	}
+	err = yaml.Unmarshal(f, &vsf)
```
Comment (Pantani) [inline pkg/ibctypes/validator_set.go:37]: unhandled error

File: x/monitoringc/keeper/verified_client_id.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+		verifiedClientID.ClientID,
+	), b)
+}
+
+// GetVerifiedClientID returns a verifiedClientID from its index
+func (k Keeper) GetVerifiedClientID(
+	ctx sdk.Context,
+	launchID uint64,
+	clientID string,
+
```
Comment (Pantani) [inline x/monitoringc/keeper/verified_client_id.go:24]:
```suggestion
```

File: x/monitoringc/keeper/verified_client_id.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	k.cdc.MustUnmarshal(b, &val)
+	return val, true
+}
+
+// RemoveVerifiedClientID removes a verifiedClientID from the store
+func (k Keeper) RemoveVerifiedClientID(
+	ctx sdk.Context,
+	launchID uint64,
+	clientID string,
+
```
Comment (Pantani) [inline x/monitoringc/keeper/verified_client_id.go:45]:
```suggestion
```

File: x/monitoringc/keeper/verified_client_id_test.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/require"
+	keepertest "github.com/tendermint/spn/testutil/keeper"
+	"github.com/tendermint/spn/testutil/nullify"
+	"github.com/tendermint/spn/x/monitoringc/keeper"
+	"github.com/tendermint/spn/x/monitoringc/types"
+)
+
+// Prevent strconv unused error
+var _ = strconv.IntSize
```
Comment (Pantani) [inline x/monitoringc/keeper/verified_client_id_test.go:16]:
```suggestion
```

File: x/monitoringc/keeper/verified_client_id_test.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+			item.ClientID,
+		)
+		require.True(t, found)
+		require.Equal(t,
+			nullify.Fill(&item),
+			nullify.Fill(&rst),
+		)
+	}
+}
+func TestVerifiedClientIDRemove(t *testing.T) {
```
Comment (Pantani) [inline x/monitoringc/keeper/verified_client_id_test.go:44]:
```suggestion

func TestVerifiedClientIDRemove(t *testing.T) {
```

File: proto/types/ibctypes.proto (hunk 2 above)
Comment (lubtd) [reply proto/types/ibctypes.proto]: Based on the thread
https://github.com/tendermint/spn/issues/454

File: pkg/ibctypes/validator_set.go (hunk 5 above)
Comment (lubtd) [reply pkg/ibctypes/validator_set.go:23]: Yes, we also track this part for refactoring. Since the logic is off-chain, this is lower priority

**Review by Pantani — APPROVED** [review]

**Pull request merged**